### PR TITLE
WIP: Add tracing configuration protos

### PR DESCRIPTION
**Pull request by douglas-reid**
This PR is a strawman PR for enabling tracing configuration for Istio. The intent
of this PR is to provide a basis for enabling CRD-driven control over tracing
parameters used throughout the mesh.

With such a CR, all proxies, Istio components, and applications themselves could receive
their tracing information from the k8s API Server (instead of having distinct configuration mechanisms).
Additionally, when fully-supported by Pilot/Envoy, such a CustomResource would allow dynamic
tuning of trace sampling rates.

I imagine that in the near future, we will want to add trace sampling rate support to
individual route rules, enabling differing (dynamic) levels of tracing per route.

This would replace the existing `enable_tracing` and `zipkin_address` configuration that
is contained in the mesh_config configmap currently in-use today. Providing support for it
would be a step towards complete CRD-driven mesh configuration.

### CODE REVIEW

**Comment by douglas-reid** [comment]
Reviewers: Interested in feedback on approach, as well as overall structure.

File: tracing/v1alpha1/tracing.proto [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+// ```yaml
+// apiVersion: tracing.istio.io/v1alpha1
+// kind: TracingConfiguration
+// metadata:
+//   name: meshwide
+// spec:
+//   traceCollector:
+//     zipkinCollector:
+//       address: zipkin:9411
+// ```
```
Comment (rshriram) [inline tracing/v1alpha1/tracing.proto]: Here is one problem: this piece is part of bootstrap config that we need in the config map for pilot agents to generate the basic Envoy config. 

CRDs imply ability to dynamically change things. We are dependent on static config maps for enabling tracing globally. Are you planning to convert the global config map into a CRD that will allow us to dynamically control the global config? This piece here would fall in place automatically

File: tracing/v1alpha1/tracing.proto [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+// and random sampling. Client-directed sampling refers to sampling based on
+// values of HTTP headers set by requesting clients. Random sampling refers to
+// the process of selecting a request for tracing based on a randomly generated
+// request ID. It is also possible to set an upper limit on the rate of all
+// trace sampling, via control of the overall sampling rate.
+message TraceSampling {
+
+    // Target percentage of requests that will be force traced if the
+    // x-client-trace-id header is set by the client of a request.
+    //
```
Comment (rshriram) [inline tracing/v1alpha1/tracing.proto]: Isn’t it 100% if that header is set?

Comment (douglas-reid) [reply tracing/v1alpha1/tracing.proto]: Not in Envoy config. I think we would probably want to set it at 100% (the envoy default), but it can be turned down.

File: tracing/v1alpha1/tracing.proto (hunk 1 above)
Comment (douglas-reid) [reply tracing/v1alpha1/tracing.proto]: Is there a piecemeal way that we could transition from the config map to CRD(s)?  Converting the entirety of the mesh and proxy config over to CRDs _feels_ like a much bigger change.

In the meantime, we could just add `TracingConfiguration` to the configmap (deprecating `zipkin_address`) right? Maybe that is the better way to get this bit of functionality in?

Comment (ayj) [reply tracing/v1alpha1/tracing.proto]: > We are dependent on static config maps for enabling tracing globally. Are you planning to convert the global config map into a CRD that will allow us to dynamically control the global config?

Mounted ConfigMaps are updated automatically (i.e. dynamically). That, plus `fsnotify`, gives us the ability to dynamically update/monitor singleton resources such as this. The sidecar injector webhook follows this approach for consuming the mesh and injection configuration.

https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#mounted-configmaps-are-updated-automatically

Comment (rshriram) [reply tracing/v1alpha1/tracing.proto]: Yes, but remember Kuat's PR on dynamic mesh config that used CRDs? We already have a well oiled system for watching CRDs, reacting to changes, etc. Having everything in CRD makes it easier for us to enforce access control, manage all of istio config as one unit instead of splitting across config maps and crds. 

I am not opposed to config maps, just opposed to having two things (crd/configmap). And pilot code does not have watches for the global mesh config. Rather than wiring up those watches all over pilot code, its much easier to piggyback on the existing crd pipeline, imo.

Comment (douglas-reid) [reply tracing/v1alpha1/tracing.proto]: I'd like the tracing config to be available to things beyond just the pilot agent. For instance, Mixer's own tracing (currently using istio.io/istio/pkg/tracing) should pull the config from the same spot, ideally. Same should be true of applications deployed in the mesh.

I guess this could be achieved in a number of ways, including potentially mutating webhooks. Do we believe that configmaps are the best way to supply this config? And does it make sense to separate tracing configuration from the rest of mesh config ( i can't imagine non-proxy components caring about proxy_admin_port or concurrency) ?

Comment (ayj) [reply tracing/v1alpha1/tracing.proto]: ConfigMaps map more naturally to singleton resources such as global mesh config in its current form. That said, I think CRD vs. ConfigMap is outside the scope of this PR. Any more selective scoping of bootstrap config applies to much more than just tracing.

Comment (ayj) [reply tracing/v1alpha1/tracing.proto]: I think for the purposes of this PR it would be useful to ignore the CRD vs. ConfigMap discussion. As @rshriram noted, this will likely change in the future as we refine how config is managed/distributed. Instead, it may be more worthwhile to consider user roles, scoping, etc. Once that's understood we can figure out the best way to map to the current and future config mgmt schemes. For example, should we distinguish configuration of the tracing sink (e.g. admin role?) from enabling the tracing feature for particular components, namespaces, services, workloads? That may require separate resource kinds with different ACLs.

Comment (douglas-reid) [reply tracing/v1alpha1/tracing.proto]: @ayj i imagine that for the vast majority of deployments there will be no need to differently ACL the collection rate from the collector config.  If I can turn tracing ON, shouldn't I also be able to select where tracing is going. And, if I can configure where tracing is going, shouldn't i be able to control the rate at which data is sent?

I do think we want to be able to configure collection rates for individual routes (as overrides).  "Turn sampling up for canary instances", or "turn sampling up for the details service" seem like reasonable use cases.

Should we schedule some time to chat quickly? I want to make sure that we head in the right direction.

Comment (ayj) [reply tracing/v1alpha1/tracing.proto]: A quick chat works for me.

Comment (rshriram) [reply tracing/v1alpha1/tracing.proto]: Given that this could apply beyond just pilot agent, I think having a separate global CRD or whatever is a better alternative.

File: tracing/v1alpha1/tracing.proto (hunk 2 above)
Comment (rshriram) [reply tracing/v1alpha1/tracing.proto]: We need to clarify this with examples. The approach I have been following is to demystify some of envoy's docs, and stick to well known terms that people understand. The relationship between the three vars is not exactly clear.

**Comment by douglas-reid** [comment]
PTAL.  After reviewing the comments here (and offline), I've tried to add some details / simplifications to the protos.

Key changes:
- adds `service_name` to the `TracingRule`. I'm the least certain about this as a mechanism to provide filtering, etc. Comments here are most welcome.
- introduces `TraceSampling.Mode` enum in attempt to make sampling configuration simple(r).
- add LightStep collector config support

File: tracing/v1alpha1/tracing.proto (hunk 1 above)
Comment (loewenstein) [reply tracing/v1alpha1/tracing.proto]: @rshriram as far as I see the pilot-agent does not use the config map at all.
E.g. the zipkin config is done via cmd flag [here](https://github.com/istio/istio/blob/master/pilot/cmd/pilot-agent/main.go#L264) and [here](https://github.com/istio/istio/blob/master/pilot/cmd/pilot-agent/main.go#L135)

In fact neither `istioctl kube-inject` nor the sidecar-injector webhook configure the pilot-agent container to mount the config map (see [here](https://github.com/istio/istio/blob/master/pilot/pkg/kube/inject/mesh.go#L129-L141) and [here](https://github.com/istio/istio/blob/master/install/kubernetes/templates/istio-sidecar-injector-configmap-release.yaml.tmpl#L83-L94))

File: tracing/v1alpha1/tracing.proto [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+// $title: Tracing V1 Alpha 1
+// $overview: Configuration affecting Istio service tracing
+
+// Provides configuration parameters for Istio tracing.
+package istio.tracing.v1alpha1;
+
+option go_package="istio.io/api/tracing/v1alpha1";
+
+// TracingRule controls Istio mesh trace collection configuration
```
Comment (geeknoid) [inline tracing/v1alpha1/tracing.proto]: I don't think I'm crazy about using the Rule suffix here. We've been using 'rule' for policy control, whereas this is not related.

File: tracing/v1alpha1/tracing.proto [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+//
+// Example configuration:
+// ```yaml
+// apiVersion: tracing.istio.io/v1alpha1
+// kind: TracingConfiguration
+// metadata:
+//   name: meshwide
+// spec:
+//   traceCollector:
+//     zipkinCollector:
```
Comment (geeknoid) [inline tracing/v1alpha1/tracing.proto]: Do we need to repeat "Collector" in this name?

File: tracing/v1alpha1/tracing.proto [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+option go_package="istio.io/api/tracing/v1alpha1";
+
+// TracingRule controls Istio mesh trace collection configuration
+// parameters. It includes the configuration for the tracing backends as well
+// as the rates at which distributed tracing data should be generated.
+//
+// Example configuration:
+// ```yaml
+// apiVersion: tracing.istio.io/v1alpha1
+// kind: TracingConfiguration
```
Comment (geeknoid) [inline tracing/v1alpha1/tracing.proto]: Doesn't match the proto name...

File: tracing/v1alpha1/tracing.proto [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+// $title: Tracing V1 Alpha 1
+// $overview: Configuration affecting Istio service tracing
+
+// Provides configuration parameters for Istio tracing.
+package istio.tracing.v1alpha1;
+
+option go_package="istio.io/api/tracing/v1alpha1";
+
+// TracingRule controls Istio mesh trace collection configuration
+// parameters. It includes the configuration for the tracing backends as well
```
Comment (geeknoid) [inline tracing/v1alpha1/tracing.proto]: "configuration to acess the tracing backends".

File: tracing/v1alpha1/tracing.proto [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+// metadata:
+//   name: meshwide
+// spec:
+//   traceCollector:
+//     zipkinCollector:
+//       address: zipkin:9411
+// ```
+//
+// WARNING: This is not currently supported by Istio.
+message TracingRule {
```
Comment (ayj) [inline tracing/v1alpha1/tracing.proto]: Usage would be simplified if we eliminated duplicate contextual information from some of the field names, e.g.

Drop `trace` prefix from fields in TracingRule
s/`trace_collector`/`collector`
s/`trace_sampling`/`sampling`

Drop `collector` suffix from fields in TracingCollector
s/`zipkin_collector`/`zipkin` 
s/`lightstep_collector`/`lightstep` 

That would give us the following

```yaml
apiVersion: tracing.istio.io/v1alpha1
kind: TracingConfiguration
metadata:
    name: meshwide
￼spec:
    service: foo-bar-service        
    collector:
￼        zipkin:
            address: zipkin:9411
            apiEndpoint: /api/v2/spans
```

File: tracing/v1alpha1/tracing.proto [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+// Configuration for a LightStep collector.
+message LightStepCollector {
+
+    // Required. Address of the LightStep service (e.g. _collector-grpc.lightstep.com:443_).
+    string host = 1;
+
+    // Required. File containing the access token for the LightStep API.
+    //
+    // Default: "/app/access_token"
+    string access_token_file = 2;
```
Comment (ayj) [inline tracing/v1alpha1/tracing.proto]: Who needs access to this token and how is the file that contains it managed? It looks like we're assuming the file is mounted via some out-of-band process not documented here (e.g. ConfigMap?).

File: tracing/v1alpha1/tracing.proto (hunk 3 above)
Comment (ayj) [reply tracing/v1alpha1/tracing.proto]: Have we formalized conventions for use of `Rule` vs. `Policy` in Istio configuration? 

```bash
$ kubectl get crd | grep -e rule -e polic
destinationpolicies.config.istio.io                       33d
destinationrules.config.istio.io                          29d
egressrules.config.istio.io                               33d
enduserauthenticationpolicyspecbindings.config.istio.io   34d
enduserauthenticationpolicyspecs.config.istio.io          34d
ingressrules.config.istio.io                              29d
routerules.config.istio.io                                33d
rules.config.istio.io                                     33d
v1alpha2routerules.config.istio.io                        34d
```

**Comment by objectiser** [comment]
@douglas-reid Like the general concept as a solution to my use case discussed [here](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/istio-users/QzNhRmiS0t4/0Ty5Mx0aAQAJ).

Just a couple of points:
* Agree with @ayj  that the configuration of the trace collector should be separate from the configuration of the sampling - for example, I wouldn't expect configuration of collector per service - although possibly you have other use cases for this?
* Would it be possible to have sampling configuration per endpoint, not just service? So for `/metrics` in my case?
  * Case 1: Service X has two endpoints, `/op1` and `/op2` - op1 is of limited interest so want minimal tracing, but op2 is critical so want 100% tracing.
  * Case 2: Sampling of all endpoints on Service X should have same sampling rate
  * Case 3: Endpoint `/xyz` should have a particular sampling rate, regardless of service

Other slightly unrelated question - if the collector can be configured to use LightStep, and others potentially in the future, does this mean there is no longer plans to consolidate the tracing data via the mixer and provide adapters to other tracing backends?

**Comment by geeknoid** [comment]
@douglas-reid This is a year old, can we close?

**Comment by douglas-reid** [comment]
Thanks for all of the good input. I think we can close this PR. I've a separate proposal doc that more closely reflects the current situation. If we need to fall back to a first class API, I can revive. Closing.
